core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1791,6 +1791,14 @@ private boolean needNewSubQuery(
         return false;
       }
       final Clause maxClause = Collections.max(clauses);
+
+      if (rel instanceof Project
+          && clauses.contains(Clause.ORDER_BY)
+          && dialect.getConformance().isSortByOrdinal()) {
+        // Cannot merge a Project that contains sort by ordinal under it.
+        return hasSortByOrdinal();
```

Comment:
To be honest, I felt uneasy about moving this up. I'm not sure if we will miss some opportunity to add a new sub query. The changes in https://github.com/apache/calcite/pull/3057 was also merged by me, and now I think that is not perfect as well, since it also suffers from this.

For `needNewSubQuery`, we should only bail out when we returns `true`. In that way, we could safely add new conditions into this method, and all the cases are orthogonal.

Reply:
Agree, we should `return true` avoid `return false` blocking other conditions.